### PR TITLE
Remove subsegments lock

### LINKED ISSUE #276: Deadlock when streaming some subsegments when parent is still open
Hi,

after an upstep to 2.9.0 we are seeing a deadlock in lock ordering, when multiple threads are adding segments to the same parent segment.

If you look at the stack traces you see the following behavior.

One thread is ending a subsegment and deciding to call `DefaultStreamingStrategy.streamSome` which takes the subsegmentsLock() to decide which it can stream.  When it finds a segment which is finished, it will try to get get lock on the parent segment to get its id[1].

But this lock is held by another thread, which tries to add a subsegment to the same parent, for that if first takes a lock on the parent segment and then it request the subsegmentLock() which is held by the first thread. [2]

We think this is a case of deadlock by lock ordering.

A possible solution might be to never take the subsegmentslock when the parent lock is not taken.  But this might influence performance.

Can you take a look?

[1]:
```
"pool-4-thread-1@7757" prio=5 tid=0x17 nid=NA waiting for monitor entry
  java.lang.Thread.State: BLOCKED
	 waiting for pool-4-thread-3@7761 to release lock on <0x1fa4> (a java.lang.Object)
	  at com.amazonaws.xray.entities.EntityImpl.getId(EntityImpl.java:233)
	  at com.amazonaws.xray.entities.SubsegmentImpl.getStreamSerializeObjectNode(SubsegmentImpl.java:132)
	  - locked <0x1fc5> (a java.lang.Object)
	  at com.amazonaws.xray.entities.SubsegmentImpl.streamSerialize(SubsegmentImpl.java:143)
	  at com.amazonaws.xray.emitters.UDPEmitter.sendSubsegment(UDPEmitter.java:97)
	  at com.amazonaws.xray.strategy.DefaultStreamingStrategy.stream(DefaultStreamingStrategy.java:120)
	  at com.amazonaws.xray.strategy.DefaultStreamingStrategy.streamSome(DefaultStreamingStrategy.java:87)
	  at com.amazonaws.xray.AWSXRayRecorder.endSubsegment(AWSXRayRecorder.java:593)
	  at com.amazonaws.xray.interceptors.TracingInterceptor.afterExecution(TracingInterceptor.java:288)
	  at software.amazon.awssdk.core.interceptor.ExecutionInterceptorChain.lambda$afterExecution$11(ExecutionInterceptorChain.java:192)
```

[2]:
```
"pool-4-thread-3@7761" prio=5 tid=0x19 nid=NA waiting
  java.lang.Thread.State: WAITING
	 blocks pool-4-thread-1@7757
	 blocks pool-4-thread-2@7758
	 blocks pool-4-thread-4@7762
	 blocks pool-4-thread-5@7763
	  at jdk.internal.misc.Unsafe.park(Unsafe.java:-1)
	  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:194)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:885)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:917)
	  at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1240)
	  at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:267)
	  at com.amazonaws.xray.entities.EntityImpl.addSubsegment(EntityImpl.java:537)
	  - locked <0x1fa4> (a java.lang.Object)
	  at com.amazonaws.xray.contexts.ThreadLocalSegmentContext.beginSubsegment(ThreadLocalSegmentContext.java:52)
	  at com.amazonaws.xray.AWSXRayRecorder.beginSubsegment(AWSXRayRecorder.java:614)
	  at com.amazonaws.xray.interceptors.TracingInterceptor.beforeExecution(TracingInterceptor.java:225)
	  at software.amazon.awssdk.core.interceptor.ExecutionInterceptorChain.lambda$beforeExecution$1(ExecutionInterceptorChain.java:62)
```

### PR DESCRIPTION
**Pull request by anuraaga**
*Issue #, if available:* Fixes #276

*Description of changes:*
Since we lock all access to the segment now, the subsegments lock is mostly not necessary. The only pattern it was needed for is user code doing something like

```java
getSubsegmentsLock().lock()
getSubsegments().add(segment) // Not addSubsegment
unlock()
```

But it's basically impossible to make code deadlock-free when returning locks to unknown code and is basically a forbidden practice in Java. I think it's best we just hope no one is doing this, and if someone is and tells us, we recommend the unfortunate code change of switching to `addSubsegment`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultStreamingStrategy.java [hunk 1]
```diff
@@ -98,14 +92,8 @@ private boolean stream(Entity entity, Emitter emitter) {
         //Gather children and in the condition they are ready to stream, add them to the streamable list.
```
Comment (steven-aerts) [inline aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultStreamingStrategy.java:92]: I think/fear a different race condition just moved 2 lines up now by removing the subsegmentLock.  And you will hit #181  again.

As the copy of `entity.getSubsegments()` is taken outside the protected region.
So it is possible that an entity is added while the subsegment list is being copied what can/will result in a `ConcurrentModificationException` being thrown.

Possible solutions I see:
* Make the collection returned by `getSubsegments` thread safe 
* Let `getSubsegments` return an (immutable) copy of the list (or provide another function which does).

**Review by steven-aerts — CHANGES_REQUESTED** [review]
I think this introduces #181 again.

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultStreamingStrategy.java (hunk 1 above)
Comment (anuraaga) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultStreamingStrategy.java:93]: Thanks for noticing this @steven-aerts - I've updated to return a copy

**Review by willarmiros — COMMENTED** [review]
I wonder if we can cook up a test to try to reproduce #276? I'm imagining something with 2 threads, forcing the deadlock with `Thread.sleep` where needed. If it's too flaky or too much work that's ok, but would be nice to have.

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/entities/EntityImpl.java [hunk 2]
```diff
@@ -530,16 +530,19 @@ public void setParentId(@Nullable String parentId) {
         }
     }
 
+    @JsonIgnore
+    @Override
+    public List<Subsegment> getSubsegmentsCopy() {
+        synchronized (lock) {
+            return new ArrayList<>(subsegments);
```
[the remark this answers is not in the document]
Comment (steven-aerts) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/entities/EntityImpl.java:537]: Personally I think it is more clear, but won't mind if the proposed behavior is kept.

**Review by steven-aerts — APPROVED** [review]
> I wonder if we can cook up a test to try to reproduce #276? I'm imagining something with 2 threads, forcing the deadlock with `Thread.sleep` where needed. If it's too flaky or too much work that's ok, but would be nice to have.

@willarmiros I tried when I created this ticket but was not able to do so.  Only saw it in live code :-( .

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/entities/EntityImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (anuraaga) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/entities/EntityImpl.java:537]: Just wrapping the subsegments won't solve the thread safety possibility, it needs to be a copy. I could also wrap the copy with Unmodifiable but I don't think we really need to for copies.

[the remark this answers is not in the document]
Comment (willarmiros) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/entities/EntityImpl.java:537]: Ok fair enough, yeah I guess doing an `UnmodifiableCollection` would still allow the user to mutate the individual subsegments.

**Review by willarmiros — APPROVED** [review]
LGTM

**Pull request merged**